### ISSUE TITLE
Autofill matching on TLDs may expose credentials.

### ISSUE THREAD

**Issue by thblt** [passff/passff#199]
If it can't match the exact domain, passff's autofill feature seems to try to match on parts of it, up to the tld.  This means that if I have a password in store for, say, `abc.com`, and visit a site called `def.com` which shows a login page, passff will autofill it with my login and password for `abc.com`, or the first .com domain (in alphabetical order) it knows.

This may accidentally expose credentials, and may even allow for specially crafted sites (eg, the whole contents a `<form>`, every link an `<input type="submit">`, input fields somehow  hidden) to randomly steal credentials from passff users. 

A website with the form fields hidden by CSS and every single link a submit button could then randomly steal credentials.  Fortunately passff apparently won't autofill a field called `url`, which somehow restricts the impact of the issue.

**Comment by tuxor1337** [comment]
Yes, an auto fill feature is a potential security issue - in any case. We can remove this feature or we can enforce exact url match which will basically render this feature useless for most users. So what's your preference?

**Comment by thblt** [comment]
Thanks for your reply.  I'll try to answer your points in order.

> Yes, an auto fill feature is a potential security issue - in any
> case. We can remove this feature or we can enforce exact URL matching
> which will basically render this feature useless for most users.

There's a third option, which is to require *at least* a full match on
the domain name in pass.  That is, allowing subdomains, but not
"superdomains".  Eg, if the pass record's name is `alpha.example.com`,
passff could match alpha.example.com, beta.alpha.example.com, but
neither example.com nor theta.example.com.

> or we can enforce exact URL matching which will basically render this
> feature useless for most users.

Maybe as an option?  An even more powerful strategy, but which would
require deeper changes in passff, would be to add a `matchdomain:`
option in the pass records themselves, with a regex or a glob or
something similar (or just a list of domains this password is for)

> even the latter option is still a security issue because we can't
> exclude that a user stores the password for foo.bar.com under the name
> bar.com, hence exposing the password to bar.com that should be visible
> to foo.bar.com only.

I understand the concern, but I can't think of an example where this
would be an issue.  There may be some of course, but for all the domains
I can think of, a password given to `something.example.com` may safely
be exposed to `example.com`.

The converse (exposing the password for bar.com to foo.bar) remains
problematic, and isn't addressed by more "third option" above.  For
example, some free/cheap webhosting services (eg, github, although
github uses a different domain, so no problem there) still host websites
with URLs like something.freewebhosting.com.  I wouldn't want my
password for freewebhosting.com exposed to roguesite.freewebhosting.com.
But if freewebhosting.com accidentally gets my password for
amazingforum.freewebhosting.com, well, if they wanted it they could have
had it easily anyway.  But a "strict match" option would address that.

I can *try* to have a look at implementing at least "full domain
matching" and "strict matching" as options.  I don't have much
experience at JavaScript, but the code seems simple enough (but if
someone else feels like doing it, I'd be more than happy to let them
handle this :)

**Comment by nwallace** [comment]
Thanks for the issue @thblt. You're right, of course, the auto-fill-and-submit feature could expose your credentials to a malicious site. However, that behavior is optional, so if you're concerned about it you can always turn it off. As a matter of fact, it's off by default, so you'd have to opt into it in the first place. So  because of that, I don't feel like this is much of an issue personally.

At the same time, our algorithm for URL matching isn't great. If we just addressed that problem it would probably avoid 90% of potential issues by itself.  What do you think?

**Comment by thblt** [comment]
Thanks for your reply, @nwallace.  

> However, that behavior is optional, so if you're concerned about it you can always turn it off.

The fact that it's opt-in is actually a good thing, yet there are numerous websites which ask you to confirm your login from time to time, usually by re-entering your password.  For these, autofill is quite useful.

> At the same time, our algorithm for URL matching isn't great. If we just addressed that problem it would probably avoid 90% of potential issues by itself. What do you think?

That's what the issue was about, indeed, I'm afraid I didn't make myself clear enough.  The issue I've encountered was that autofilling didn't know when to give up, so to say: On a `.com` domain absent from the pass db, it would try to fill with credentials from the first `.com` domain available.  Eg, if I'm browsing `mallory.com`, pass would try to fill in my credentials for `alice.com`, because `.com` is common to the two domains.   IMHO, the URL matching algorithm for autofill should:

 - provide a "strict" option, where eg the record `that.example.com.gpg`  would match the domain `that.example.com`, but neither its subdomains `this.that.example.com` nor its superdomains (`example.com`) 
 - in non-strict mode, give up if it can't match at least domain+tld (`example.com`)

**Comment by nwallace** [comment]
@thblt That's a good idea -- to have a configuration option to control how strict the domain name matching is. I've got [a branch](https://github.com/passff/passff/tree/beautification/rewrite) going (originally intended to just refactor & simplify the codebase) where I incorporated [fuzzaldrin-plus](https://github.com/jeancroy/fuzz-aldrin-plus) for the URL matching. This version works similarly to what you propose, and I could pretty easily add a configuration option to use stricter matching.

Since this issue is more about the URL-matching algorithm, I'm going to close this issue and direct any further discussion about the algorithm to #78
